File: backend/src/main/java/com/chatapp/server/WebSocketBridge.java
package com.chatapp.server;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.java_websocket.WebSocket;
import org.java_websocket.handshake.ClientHandshake;
import org.java_websocket.server.WebSocketServer;

/**
 * WebSocket Bridge Server that bridges WebSocket connections to the TCP chat server
 */
public class WebSocketBridge extends WebSocketServer {
    private static final int WS_PORT = 8082;
    private static final String JAVA_SERVER_HOST = "localhost";
    private static final int JAVA_SERVER_PORT = 8081;
    
    private final ConcurrentHashMap<WebSocket, Socket> wsToTcpMap;
    private final ConcurrentHashMap<WebSocket, BufferedReader> wsToReaderMap;
    private final ConcurrentHashMap<WebSocket, PrintWriter> wsToWriterMap;
    private final ExecutorService executorService;

    public WebSocketBridge() {
        super(new InetSocketAddress(WS_PORT));
        this.wsToTcpMap = new ConcurrentHashMap<>();
        this.wsToReaderMap = new ConcurrentHashMap<>();
        this.wsToWriterMap = new ConcurrentHashMap<>();
        this.executorService = Executors.newCachedThreadPool();
        
        System.out.println("╔════════════════════════════════════════╗");
        System.out.println("║   WebSocket Bridge Server Started      ║");
        System.out.println("╚════════════════════════════════════════╝");
        System.out.println("📡 WebSocket Server: ws://localhost:" + WS_PORT);
        System.out.println("🔌 Java Server: " + JAVA_SERVER_HOST + ":" + JAVA_SERVER_PORT);
        System.out.println("⏳ Waiting for connections...\n");
    }

    @Override
    public void onOpen(WebSocket conn, ClientHandshake handshake) {
        System.out.println("🔌 New WebSocket client connected: " + conn.getRemoteSocketAddress());
        
        // Create TCP connection to Java server in a separate thread
        executorService.submit(() -> connectToTcpServer(conn));
    }

    private void connectToTcpServer(WebSocket ws) {
        try {
            Socket tcpSocket = new Socket(JAVA_SERVER_HOST, JAVA_SERVER_PORT);
            BufferedReader reader = new BufferedReader(new InputStreamReader(tcpSocket.getInputStream()));
            PrintWriter writer = new PrintWriter(tcpSocket.getOutputStream(), true);
            
            wsToTcpMap.put(ws, tcpSocket);
            wsToReaderMap.put(ws, reader);
            wsToWriterMap.put(ws, writer);
            
            System.out.println("✅ Connected to Java server for WebSocket: " + ws.getRemoteSocketAddress());
            
            // Start listening for messages from TCP server
            executorService.submit(() -> listenToTcpServer(ws, reader));
            
        } catch (IOException e) {
            System.err.println("❌ Failed to connect to TCP server: " + e.getMessage());
            ws.close();
        }
    }

    private void listenToTcpServer(WebSocket ws, BufferedReader reader) {
        try {
            String message;
            while ((message = reader.readLine()) != null && ws.isOpen()) {
                System.out.println("📥 TCP → WS: " + message);
                ws.send(message);
            }
        } catch (IOException e) {
            if (ws.isOpen()) {
                System.err.println("❌ Error reading from TCP server: " + e.getMessage());
                ws.close();
            }
        }
    }

    @Override
    public void onMessage(WebSocket conn, String message) {
        try {
            System.out.println("📤 WS → TCP: " + message);
            PrintWriter writer = wsToWriterMap.get(conn);
            if (writer != null) {
                writer.println(message);
            } else {
                System.err.println("❌ No TCP connection found for WebSocket");
            }
        } catch (Exception e) {
            System.err.println("❌ Error forwarding message to TCP: " + e.getMessage());
        }
    }

    @Override
    public void onClose(WebSocket conn, int code, String reason, boolean remote) {
        System.out.println("👋 WebSocket client disconnected: " + conn.getRemoteSocketAddress());
        cleanupConnection(conn);
    }

    @Override
    public void onError(WebSocket conn, Exception ex) {
        System.err.println("❌ WebSocket error: " + ex.getMessage());
        if (conn != null) {
            cleanupConnection(conn);
        }
    }

    private void cleanupConnection(WebSocket ws) {
        try {
            // Close TCP connection
            Socket tcpSocket = wsToTcpMap.remove(ws);
            if (tcpSocket != null && !tcpSocket.isClosed()) {
                tcpSocket.close();
            }
            
            // Close reader
            BufferedReader reader = wsToReaderMap.remove(ws);
            if (reader != null) {
                reader.close();
            }
            
            // Close writer
            PrintWriter writer = wsToWriterMap.remove(ws);
            if (writer != null) {
                writer.close();
            }
            
            System.out.println("🔌 TCP connection cleaned up for WebSocket");
        } catch (IOException e) {
            System.err.println("❌ Error cleaning up connection: " + e.getMessage());
        }
    }

    @Override
    public void onStart() {
        System.out.println("🚀 WebSocket server started successfully!");
        setConnectionLostTimeout(0);
        setConnectionLostTimeout(100);
    }

    public void shutdownBridge() {
        System.out.println("\n🛑 Shutting down WebSocket bridge...");
        
        // Clean up all connections
        wsToTcpMap.keySet().forEach(this::cleanupConnection);
        
        // Shutdown executor service
        executorService.shutdown();
        
        try {
            this.stop();
            System.out.println("✅ WebSocket server closed");
        } catch (InterruptedException e) {
            System.err.println("❌ Error stopping WebSocket server: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        WebSocketBridge server = new WebSocketBridge();
        server.start();

        // Graceful shutdown hook
        Runtime.getRuntime().addShutdownHook(new Thread(server::shutdownBridge));

        // Keep the main thread alive
        try {
            Thread.currentThread().join();
        } catch (InterruptedException e) {
            System.err.println("Main thread interrupted");
        }
    }
}

File: backend/websocket-bridge.js
// WebSocket to TCP Socket Bridge
// This server bridges WebSocket connections from the browser to the Java TCP server

const WebSocket = require('ws');
const net = require('net');

const WS_PORT = 8082;
const JAVA_SERVER_HOST = 'localhost';
const JAVA_SERVER_PORT = 8081;

const wss = new WebSocket.Server({ port: WS_PORT });

console.log('╔════════════════════════════════════════╗');
console.log('║   WebSocket Bridge Server Started      ║');
console.log('╚════════════════════════════════════════╝');
console.log(`📡 WebSocket Server: ws://localhost:${WS_PORT}`);
console.log(`🔌 Java Server: ${JAVA_SERVER_HOST}:${JAVA_SERVER_PORT}`);
console.log('⏳ Waiting for connections...\n');

wss.on('connection', (ws) => {
  console.log('🔌 New WebSocket client connected');

  // Create TCP connection to Java server
  const tcpClient = new net.Socket();

  tcpClient.connect(JAVA_SERVER_PORT, JAVA_SERVER_HOST, () => {
    console.log('✅ Connected to Java server');
  });

  // Forward WebSocket messages to TCP socket
  ws.on('message', (message) => {
    try {
      const data = message.toString();
      console.log('📤 WS → TCP:', data);
      tcpClient.write(data + '\n'); // Java server expects newline-delimited messages
    } catch (error) {
      console.error('❌ Error forwarding to TCP:', error);
    }
  });

  // Forward TCP socket messages to WebSocket
  tcpClient.on('data', (data) => {
    try {
      const messages = data.toString().split('\n').filter(msg => msg.trim());
      messages.forEach(msg => {
        console.log('📥 TCP → WS:', msg);
        ws.send(msg);
      });
    } catch (error) {
      console.error('❌ Error forwarding to WebSocket:', error);
    }
  });

  // Handle disconnections
  ws.on('close', () => {
    console.log('👋 WebSocket client disconnected');
    tcpClient.end();
  });

  tcpClient.on('close', () => {
    console.log('🔌 TCP connection closed');
    ws.close();
  });

  tcpClient.on('error', (error) => {
    console.error('❌ TCP error:', error.message);
    ws.close();
  });

  ws.on('error', (error) => {
    console.error('❌ WebSocket error:', error.message);
    tcpClient.end();
  });
});

wss.on('error', (error) => {
  console.error('❌ WebSocket Server error:', error);
});

// Graceful shutdown
process.on('SIGINT', () => {
  console.log('\n🛑 Shutting down WebSocket bridge...');
  wss.close(() => {
    console.log('✅ WebSocket server closed');
    process.exit(0);
  });
});
